lib/OpenQA/WebAPI/Controller/API/V1/JobTemplate.pm

```diff
@@ -200,7 +207,195 @@ sub schedules {
+=cut
+
+sub update {
+    my $self = shift;
+
+    my $yaml   = {};
+    my $errors = [];
+    try {
+        # No objects (aka SafeYAML)
+        $YAML::XS::LoadBlessed = 0;
+        $yaml                  = YAML::XS::Load($self->param('template'));
+        $errors                = $self->app->validate_yaml($yaml, $self->app->log->level eq 'debug');
+    }
+    catch {
+        # Push the exception to the list of errors without the trailing new line
+        push @$errors, substr($_, 0, -1);
+    };
+
+    if (@$errors) {
+        $self->app->log->error(@$errors);
+        $self->respond_to(json => {json => {error => \@$errors}, status => 400},);
+        return;
+    }
+
+    my $job_groups    = $self->schema->resultset("JobGroups");
+    my $job_templates = $self->schema->resultset("JobTemplates");
+    my $machines      = $self->schema->resultset("Machines");
+    my $test_suites   = $self->schema->resultset("TestSuites");
+    my $products      = $self->schema->resultset("Products");
+    foreach my $group (keys %{$yaml}) {
+        my $json = {};
+
+        try {
+            $self->schema->txn_do(
+                sub {
+                    my $id = $job_groups->find_or_create({name => $group}, {select => [qw(id name)]})->id;
+                    $json->{id} = $id;
+
+                    # Find existing templates to update or drop
+                    my $templates = $job_templates->search({group_id => $id}, {order_by => 'me.test_suite_id'});
+                    while (my $template = $templates->next) {
+                        my $arch         = $template->product->arch;
+                        my $product_name = $template->product->name;
+
+                        if (my $product = $yaml->{$group}->{architectures}->{$arch}{$product_name}) {
+                            my $found = 0;
+                            foreach my $test_suite (@{$product}) {
+                                if (ref $test_suite eq 'HASH') {
+                                    foreach my $name (keys %$test_suite) {
+                                        next unless $name eq $template->test_suite->name;
+                                        $found = 1;
+
+                                        my $attr = $test_suite->{$name};
+                                        if (%$attr) {
+                                            if ($attr->{machine}) {
+                                                my $machine = $machines->find({name => $attr->{machine}});
+                                                die "Machine '$attr->{machine}' is invalid\n" unless $machine;
+                                                $template->machine_id($machine->id);
+                                            }
+                                            if ($attr->{priority}) {
+                                                $template->prio($attr->{priority});
+                                            }
+                                            $template->update;
```

Comment:
When simply re-saving job templates from 'Functional: Server' from OSD I get:

```
DBIx::Class::Storage::DBI::_dbh_execute(): DBI Exception: DBD::Pg::st execute failed: FEHLER: doppelter Schlüsselwert verletzt Unique-Constraint »job_templates_product_id_machine_id_test_suite_id« DETAIL: Schlüssel »(product_id, machine_id, test_suite_id)=(385, 86, 1654)« existiert bereits. [for Statement "UPDATE job_templates SET machine_id = ?, t_updated = ? WHERE ( id = ? )" with ParamValues: 1='86', 2='2019-04-25 11:26:33', 3='12932'] at /hdd/openqa-devel/repos/openQA/script/../lib/OpenQA/WebAPI/Controller/API/V1/JobTemplate.pm line 287
```
So in this update a unique constraint might be violated. The same happens when re-saving job templates for the YaST group. But for instance 'Migration: Server' works.

I haven't looked into this problem to I can not tell you what's the cause (it could also be the export). By the way, my OSD import is from 4 month ago so maybe it is not reproducible with current data.

Reply:
Just for the record, this would be resolved with the commit from https://github.com/os-autoinst/openQA/pull/2056 (which hopefully does not cause other problems?).

Reply:
I pulled your awesome fixes into the branch. Much nicer especially with the nifty deleting!